init/fluent_bit_init_process.go

```diff
@@ -0,0 +1,401 @@
+		}
+		// if this env var's value is a path of our built-in config file, process is derectly
+		if matched_file {
+			processConfigFile(envValue)
+		}
+	}
+}
+
+func processConfigFile(path string) {
+
+	content_b, err := ioutil.ReadFile(path)
+	if err != nil {
+		logrus.Errorln(err)
+		logrus.Fatalf("[FluentBit Init Process] Cannot open file: %s\n", path)
+	}
+
+	content := string(content_b)
+
+	if strings.Contains(content, "[PARSER]") {
+		// this is a parser config file, change command
+		updateCommand(path)
+	} else {
+		// this is not a parser config file. @INCLUDE
+		writeInclude(path, mainConfigFile)
+	}
+}
+
+func getS3ConfigFile(arn string) {
+
+	// Preparation for downloading S3 config files
+	if !exists3Client {
+		createS3Client()
+	}
+
+	// e.g. "arn:aws:s3:::ygloa-bucket/s3_parser.conf"
+	arnBucketFile := arn[13:]
+	bucketAndFile := strings.SplitN(arnBucketFile, "/", 2)
+	if len(bucketAndFile) != 2 {
+		logrus.Fatalf("[FluentBit Init Process] Unrecognizable arn: %s\n", arn)
+	}
+
+	bucketName := bucketAndFile[0]
+	s3FilePath := bucketAndFile[1]
+
+	// get bucket region
+	input := &s3.GetBucketLocationInput{
+		Bucket: aws.String(bucketName),
+	}
+
+	output, err := s3Client.GetBucketLocation(input)
+	if err != nil {
+		logrus.Errorln(err)
+		logrus.Fatalf("[FluentBit Init Process] Cannot get bucket region of %s + %s", bucketName, s3FilePath)
+	}
+
+	bucketRegion := aws.StringValue(output.LocationConstraint)
+	// Buckets in Region us-east-1 have a LocationConstraint of null
+	// https://docs.aws.amazon.com/sdk-for-go/api/service/s3/#GetBucketLocationOutput
+	if bucketRegion == "" {
+		bucketRegion = "us-east-1"
+	}
+
+	// create a downloader
+	s3Downloader := createS3Downloader(bucketRegion)
+
+	// download file from S3 and store in the folder "/init/fluent-bit-init-s3-files/"
+	downloadS3ConfigFile(s3Downloader, s3FilePath, bucketName, s3FileFolderPath)
+
+}
+
+// create a S3 client as the global S3 client for reuse
+func createS3Client() {
+
+	region := "us-east-1"
+	if metadataReigon != "" {
+		region = metadataReigon
+	}
+
+	s3Client = s3.New(session.Must(session.NewSession(&aws.Config{
+		// if not specify region here, missingregion error will raise when get bucket location
+		Region: aws.String(region),
+	})))
+
+	exists3Client = true
+}
+
+func createS3Downloader(bucketRegion string) S3Downloader {
+	sess, err := session.NewSession(&aws.Config{
+		Region: aws.String(bucketRegion)},
+	)
+	if err != nil {
+		logrus.Errorln(err)
+		logrus.Fatalln("[FluentBit Init Process] Cannot creat a new session")
+	}
+
+	// need to specify session region!
+	s3Downloader := s3manager.NewDownloader(sess)
+	return s3Downloader
+}
+
+func downloadS3ConfigFile(s3Downloader S3Downloader, s3FilePath, bucketName, s3FileFolder string) {
+
+	s3FileName := strings.SplitN(s3FilePath, "/", -1)
+	fileFromS3 := createFile(s3FileFolder+s3FileName[len(s3FileName)-1], false)
+	defer fileFromS3.Close()
+
+	_, err := s3Downloader.Download(fileFromS3,
```

Comment:
Do we need to consider retry is download from S3 fail for once?

Reply:
Check the docs, I think the downloader may have some setting to auto-add retries.

Reply:
Already modified. Thanks!